pkg/controller/dnsrecord/actuator.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dnsrecord
+
+import (
+	"context"
+	"fmt"
+
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	"github.com/gardener/gardener/extensions/pkg/controller/dnsrecord"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	extensionsv1alpha1helper "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1/helper"
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener-extension-provider-aws/pkg/aws"
+	awsclient "github.com/gardener/gardener-extension-provider-aws/pkg/aws/client"
+)
+
+type actuator struct {
+	client client.Client
+	logger logr.Logger
+}
+
+func NewActuator(logger logr.Logger) dnsrecord.Actuator {
+	return &actuator{
+		logger: logger.WithName("aws-dnsrecord-actuator"),
+	}
+}
+
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+// Reconcile reconciles the DNSRecord.
+func (a *actuator) Reconcile(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster) error {
```

Comment:
One thing I noticed during my tests:
The controller uses the default workqueue rate limiter, which means failures are retried very quickly.
E.g. when specifying the wrong zone ID, this will result in a bunch of requests in a few seconds.

We might want to
- a) use a rate limiter with different qps/baseDelay settings or
- b) set `RetryAfter` (or influence it somehow from the actuator) or
- c) try to detect such errors (e.g. reusing error code detection early on here in the actuator) and only retry a few times
- or a combination of those options

in order to not immediately exhaust the accounts rate limits on such configuration issues.

Reply:
OK, good point. I think b) is easiest since we have `ReconcileErr` to override the `RequeueAfter` value returned by the reconciler. We should perhaps set this to e.g. 30 seconds for all errors that are coming from the route53 API, so no sophisticated error detection is needed? 

Regarding "only retry a few times" this would mean saving the number of retries in the status. I would avoid it unless you think it's really important.

Reply:
> OK, good point. I think b) is easiest since we have `ReconcileErr` to override the `RequeueAfter` value returned by the reconciler. We should perhaps set this to e.g. 30 seconds for all errors that are coming from the route53 API, so no sophisticated error detection is needed?

Hmmm, if we use a constant backoff of `30s` this might even make more calls in the long run that an exponential backoff, right?
But I'm not too sure about what's the best option here. We can take this to our in-person discussion tomorrow.

> Regarding "only retry a few times" this would mean saving the number of retries in the status. I would avoid it unless you think it's really important.

I can't really judge as of now, as I don't have enough hands-on experience with those things. Maybe @MartinWeindel can already give an assessment based on his experiences.
But in any case, we can still introduce such mechanism later on, if it becomes necessary.

Reply:
Implemented and tested it with a hardcoded requeueAfter of 30 seconds and a wrongly configured zone. The actual requests happen more often than 30 seconds in the beginning because gardenlet also retries. After gardenlet stops retrying, they continue happening every 30 seconds. Perhaps the value should be increased to 1 minute? I don't think it should be made configurable though, it's just a precaution against rate limits.

Reply:
> Hmmm, if we use a constant backoff of 30s this might even make more calls in the long run that an exponential backoff, right?

Yes, but I think rate limits are per unit of time, so this doesn't really matter? The important thing is not to have too many requests quickly in succession, especially if many shoots are reconciled at the same time with a wrong configuration. I am not really sure either, but as I said I would not start with some elaborate mechanism at this point just to mitigate this issue.

BTW, I think gardenlet retries mentioned above could be prevented by correctly classifying such errors as user errors. Right now this is not the case, so I am seeing more requests than optimal. I guess 30 seconds is perhaps ok in this case.

Reply:
When thinking more about it, I think the one practical way to deal with the issue would be to combine b) and c) - detect configuration errors if at all possible (e.g. the error code `NoSuchHostedZone` can't be anything else but a configuration error), and for such errors return a really long `RequeueAfter`, e.g. 1 hour. For all other route53 errors, return e.g. 30 seconds or simply propagate the error as is.
We could also recognize rate limit errors (I believe this is the `ThrottlingException` code), and `RequeueAfter` 1 hour here as well (on errors), since if we are throttled it's obviously pointless to retry more often.

Reply:
True, all good what you say. 
However, I think it's enough to save these ideas for later, once we actually have the problems. The current approach is not the nicest, but is probably already enough.

Reply:
We discussed the issue and decided not to do any further changes. Instead, we should rely on gardener properly categorizing these errors as user errors via https://github.com/gardener/gardener/blob/677eaffc5acdf78d19f09566945f4e91d97da742/pkg/apis/core/v1beta1/helper/errors.go#L56-L64 (the regexes should be extended appropriately with the corresponding route53 errors, as well as those from other providers). In this case, the shoot will rather quickly become `Failed`, and our extension controller will stop reconciling its resources, which will prevent further requests to the API. In essence, we already have an appropriate mechanism to deal with this, and we should use it instead of introducing an a different one.